### DISCUSSION TITLE
Use immutable data structures

### DISCUSSION THREAD

**Discussion by 6543** [woodpecker-ci/woodpecker#3165] [General]
well there are third-partys ...

https://github.com/benbjohnson/immutable
https://github.com/fishy/go-immutable
...
but if they work for us, i dont know. as we have to de-/serialize from/to json inbetween

but it would bring us 1. performance and 2. more savety to alter not to be altered data

**Comment by 6543** [comment]
e.g. https://github.com/woodpecker-ci/woodpecker/pull/3135#discussion_r1447293496

**Comment by lafriks** [comment]
Why do we need them?

**Reply by 6543** [reply]
1. performance
2. type savety
3. security

**Reply by lafriks** [reply]
More context would probably be needed but...
1. Immutable will have additional cost (performance/memory usage will degrade as it makes copies of lists/maps)
2. We already have type safety except only in some cases but I don't think in these cases immutable will help anyway.
